race-buffer/model/RaceBuffer.tla

```diff
@@ -6,111 +6,93 @@ CONSTANTS
 
   NumWrites  (* Number of writes to execute *)
 
-ASSUME Assumption == BufCapacity \in (Nat \ 0..2) (* Buffer capacity is at least 3 *)
+ASSUME Assumption == BufCapacity \in Nat\*(Nat \ 0..2) (* Buffer capacity is at least 3 *)
 
 -----------------------------------------------------------------------------
 
 VARIABLES 
   wpc, (* Writer's program counter *)
   buf, (* Backing storage *)
-  wcurs, (* Write cursor - Equal to number of entries already written *)
-
-  writes, (* Number of entries written (used to limit state space) *)
+  wcurs, (* Write cursor - Points to next write position *)
+  owcurs, (* Overwrite cursor - Marks where space is allocated for the writer to write freely *)
 
   rpc, (* Reader's program counter *)
   rbuf, (* Buffer of entries already read *)
   bufSnap, (* Intermediate buffer used to store a snapshot of entries being read *)
-  rcurs, (* Read cursor - Equal to number of entries already read*)
-  wcursSnap (* Value of wcurs observed at beginning of read *)
+  rcurs, (* Read cursor - Points to next read position *)
+  wcursSnap, (* Value of write cursor observed at beginning of read *)
+  storedPrefix, (* Prefix to be written to read buffer once suffix is read *)
+  processingIndex (* Index of next entry in bufSnap to process *)
 
 -----------------------------------------------------------------------------
 
 (* Writer helpers *)
 
 (* Construct an entry structure - Index field is only used to check *)
-(* ordering in the read buffer of non-nil entries. *)
+(* ordering in the reader's output buffer *)
 Entry(i, t) == [index |-> i, type |-> t]
 
-(* Get index in backing storage based on entry index *)
+(* Get index in backing storage based on given cursor *)
 BufIndex(i) == i % BufCapacity
 
-(* Get entry in storage position of given index *)
+(* Get entry in storage position of given cursor *)
 Read(i) == buf[BufIndex(i)]
 
 (* Write a single entry of given type at given index *)
 Write(i, t) == buf' = (BufIndex(i) :> Entry(i,t)) @@ buf
 
-(* Write the appropriately typed non-nil entry at the given index *)
+(* Either write a single entry or the prefix of a double entry *)
+WriteNewEntry(i) == 
+  \/ Write(i, "SINGLE")
+  \/ Write(i, "PREFIX")
+
+(* Write the appropriately typed entry at the given index *)
 WriteEntry(i) == 
-  /\ IF Read(i - 1).type = "PREFIX"
-     THEN Write(i, "SUFFIX")
-     ELSE \/ Write(i, "SINGLE")
-          \/ Write(i, "PREFIX")
-  /\ writes' = writes + 1
+  IF i = 0 
+  THEN WriteNewEntry(i) 
+  ELSE IF Read(i - 1).type = "PREFIX"
+       THEN Write(i, "SUFFIX")
+       ELSE WriteNewEntry(i)
```

Comment:
Since what gets written depends on what is read, we should allow these to be schedule separately; I don't think we can consider this an atomic operation.

Reply:
Or maybe it's ok, as this is essentially reprsenting state in the writer process?

Reply:
The Read(i) operator is just a shorthand to access the array. In practice, it is the user's responsibility to write the suffix they want, as there is no special treatment of suffixes from the writer side, except that a suffix is assumed to be right after a prefix when overwriting. Here I am just labeling the data as a suffix so the reader's output buffer can be checked for consistency.